transport/http2_server.go

```diff
@@ -143,6 +146,10 @@ func (t *http2Server) operateHeaders(hDec *hpackDecoder, s *Stream, frame header
 			hDec.state = decodeState{}
 		}
 	}()
+	if goaway {
+		//Stop creating streams on this transport
+		return nil
+	}
```

Comment:
On Thu, Apr 30, 2015 at 9:45 AM, Gian Biondi <notifications@github.com>
wrote:

> In transport/http2_server.go
> <https://github.com/grpc/grpc-go/pull/180#discussion_r29446798>:
>
> > @@ -143,6 +146,10 @@ func (t *http2Server) operateHeaders(hDec *hpackDecoder, s *Stream, frame header
> >  			hDec.state = decodeState{}
> >  		}
> >  	}()
> > +	if goaway {
> > +		//Stop creating streams on this transport
> > +		return nil
> > +	}
>
> This part I'm a little unclear on still. Can you point me to some
> documentation to explain this more?
>
This is basically the requirement of HPACK instead of gRPC. You can search
HPACK IETF draft to see whether you can find something useful. HPACK uses
huffman coding to compress http headers. Thus to be able to encode&decode a
headers frame, the states of huffman coding on sender and receiver have to
be consistent. When a receiver receives a headers frame, it means the
sender's huffman coding state has covered all the sent headers. In order to
keep consistency, the receiver has to process this headers frame and update
its own huffman coding state regardless.

Actually, I would strongly recommend writing a test case in
transport_test.go to make sure your pull request works as expected. But I
am not sure whether this is a mission impossible for you because
transport_test.go is not well constructed and hard to understand.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/180/files#r29446798>.
>